components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,67 +85,105 @@ public AbstractHTTPFunction() {
-            JSONObject json = null;
-            int responseCode;
             String outcome;
             String endpointURL = null;
 
-            if (request.getURI() != null) {
-                endpointURL = request.getURI().toString();
-            }
+            HttpUriRequest request;
+            try {
+                if (authConfigModel != null) {
+                    AuthConfig authConfig = AuthConfigFactory.getAuthConfig(authConfigModel, context, asyncReturn);
+                    request = authConfig.applyAuth(clientRequest, authConfigModel);
+                } else {
+                    request = clientRequest;
+                }
+
+                if (request.getURI() != null) {
+                    endpointURL = request.getURI().toString();
+                }
 
-            if (!isValidRequestDomain(request.getURI())) {
-                outcome = Constants.OUTCOME_FAIL;
-                LOG.error("Provided Url does not contain a allowed domain. Invalid Url: " + endpointURL);
-                asyncReturn.accept(context, Collections.emptyMap(), outcome);
-                return;
+                if (!isValidRequestDomain(request.getURI())) {
+                    LOG.error("Provided Url does not contain a allowed domain. Invalid Url: " + endpointURL);
+                    asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+                } else {
+                    Pair<String, JSONObject> result = executeRequestWithRetries
+                            (request, endpointURL, maxRequestAttemptsForAPIEndpointTimeout);
+                    outcome = result.getLeft();
+                    JSONObject json = result.getRight();
+
+                    asyncReturn.accept(context, json != null ? json : Collections.emptyMap(), outcome);
+                }
+            } catch (SecretManagementException e) {
+                LOG.error("Error while resolving secrets.", e);
+                asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
+            } catch (Exception e) {
+                LOG.error("Error while applying authentication to the request.", e);
+                asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
             }
+        });
+        JsGraphBuilder.addLongWaitProcess(asyncProcess, eventHandlers);
+    }
 
+    /**
+     * Execute the request with retries.
+     *
+     * @param request     HttpUriRequest.
+     * @param endpointURL Endpoint URL.
+     * @param maxRetries  Maximum number of retries.
+     * @return Pair of outcome and json.
+     */
+    private Pair<String, JSONObject> executeRequestWithRetries(HttpUriRequest request, String endpointURL, int maxRetries) {
+        JSONObject json = null;
+        String outcome = Constants.OUTCOME_FAIL;
+        int attempts = 0;
+
+        while (attempts < maxRetries) {
+            attempts++;
             try (CloseableHttpResponse response = client.execute(request)) {
```

Comment:
where do we set the timeout for the http request call

Reply:
https://github.com/wso2-extensions/identity-conditional-auth-functions/blob/c29ba564cad40717216931e04224f3e60b3d431f/components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java#L77-L84